File: client-app/lib/core/tool_chain_partition.dart
import 'package:gen/domain/entities/message.dart';

class ToolSegmentIndices {
  const ToolSegmentIndices({
    required this.leadIndex,
    required this.toolStart,
    required this.toolEnd,
  });

  final int leadIndex;
  final int toolStart;
  final int toolEnd;
}

class ToolChainIndices {
  const ToolChainIndices({
    required this.segments,
    this.finalAssistantIndex,
  });

  final List<ToolSegmentIndices> segments;
  final int? finalAssistantIndex;
}

sealed class MessagePartition {}

class PartitionSingle extends MessagePartition {
  PartitionSingle(this.index);
  final int index;
}

class PartitionToolChain extends MessagePartition {
  PartitionToolChain(this.chain);
  final ToolChainIndices chain;
}

bool _assistantHasToolCalls(Message m) {
  if (m.role != MessageRole.assistant) {
    return false;
  }

  return m.toolCallsJson != null && m.toolCallsJson!.trim().isNotEmpty;
}

class _ConsumeResult {
  const _ConsumeResult(this.chain, this.nextIndex);
  final ToolChainIndices chain;
  final int nextIndex;
}

_ConsumeResult? _consumeToolChain(List<Message> msgs, int start) {
  final n = msgs.length;
  var cur = start;
  final segs = <ToolSegmentIndices>[];

  while (true) {
    if (cur >= n || !_assistantHasToolCalls(msgs[cur])) {
      if (segs.isEmpty) {
        return null;
      }

      return _ConsumeResult(ToolChainIndices(segments: segs), cur);
    }

    var te = cur + 1;
    while (te < n && msgs[te].role == MessageRole.tool) {
      te++;
    }

    if (te == cur + 1) {
      if (segs.isEmpty) {
        return null;
      }
      return _ConsumeResult(ToolChainIndices(segments: segs), cur);
    }

    segs.add(ToolSegmentIndices(leadIndex: cur, toolStart: cur + 1, toolEnd: te - 1));

    if (te >= n) {
      return _ConsumeResult(ToolChainIndices(segments: segs), n);
    }

    final nxt = msgs[te];
    if (nxt.role != MessageRole.assistant) {
      return _ConsumeResult(ToolChainIndices(segments: segs), te);
    }

    if (_assistantHasToolCalls(nxt)) {
      cur = te;
      continue;
    }

    return _ConsumeResult(
      ToolChainIndices(segments: segs, finalAssistantIndex: te),
      te + 1,
    );
  }
}

List<MessagePartition> partitionMessagesForToolChainUi(List<Message> msgs) {
  final n = msgs.length;
  final out = <MessagePartition>[];
  var i = 0;
  while (i < n) {
    final m = msgs[i];
    if (!_assistantHasToolCalls(m)) {
      out.add(PartitionSingle(i));
      i++;
      continue;
    }

    final consumed = _consumeToolChain(msgs, i);
    if (consumed == null) {
      out.add(PartitionSingle(i));
      i++;
      continue;
    }

    out.add(PartitionToolChain(consumed.chain));
    i = consumed.nextIndex;
  }

  return out;
}

File: internal/chatui/tool_chain_partition.go
package chatui

import (
	"strings"

	"github.com/magomedcoder/gen/internal/domain"
)

type ToolSegment struct {
	LeadIndex int
	ToolStart int
	ToolEnd   int
}

type ToolChainGroup struct {
	Segments          []ToolSegment
	FinalAssistantIdx *int
}

type PartitionElement struct {
	SingleIndex *int
	Chain       *ToolChainGroup
}

func assistantHasToolCalls(m *domain.Message) bool {
	return m != nil &&
		m.Role == domain.MessageRoleAssistant &&
		strings.TrimSpace(m.ToolCallsJSON) != ""
}
func PartitionMessagesForToolChainUI(msgs []*domain.Message) []PartitionElement {
	n := len(msgs)
	var out []PartitionElement
	i := 0
	for i < n {
		m := msgs[i]
		if m == nil {
			out = append(out, PartitionElement{SingleIndex: intPtr(i)})
			i++
			continue
		}

		if !assistantHasToolCalls(m) {
			out = append(out, PartitionElement{SingleIndex: intPtr(i)})
			i++
			continue
		}

		group, next := consumeToolChain(msgs, i)
		if group == nil {
			out = append(out, PartitionElement{SingleIndex: intPtr(i)})
			i = next
			continue
		}

		out = append(out, PartitionElement{Chain: group})
		i = next
	}

	return out
}

func intPtr(v int) *int {
	return &v
}

func consumeToolChain(msgs []*domain.Message, start int) (group *ToolChainGroup, nextIndex int) {
	n := len(msgs)
	cur := start
	var segs []ToolSegment
	for {
		if cur >= n || msgs[cur] == nil || !assistantHasToolCalls(msgs[cur]) {
			if len(segs) == 0 {
				return nil, start + 1
			}
			return &ToolChainGroup{
				Segments:          segs,
				FinalAssistantIdx: nil,
			}, cur
		}

		te := cur + 1
		for te < n && msgs[te] != nil && msgs[te].Role == domain.MessageRoleTool {
			te++
		}

		if te == cur+1 {
			if len(segs) == 0 {
				return nil, cur + 1
			}
			return &ToolChainGroup{
				Segments:          segs,
				FinalAssistantIdx: nil,
			}, cur
		}

		segs = append(segs, ToolSegment{
			LeadIndex: cur,
			ToolStart: cur + 1,
			ToolEnd:   te - 1,
		})

		if te >= n {
			return &ToolChainGroup{
				Segments:          segs,
				FinalAssistantIdx: nil,
			}, n
		}

		nxt := msgs[te]
		if nxt == nil || nxt.Role != domain.MessageRoleAssistant {
			return &ToolChainGroup{Segments: segs, FinalAssistantIdx: nil}, te
		}

		if assistantHasToolCalls(nxt) {
			cur = te
			continue
		}

		fi := te

		return &ToolChainGroup{
			Segments:          segs,
			FinalAssistantIdx: &fi,
		}, te + 1
	}
}
